Generalized lightcurves

matteobachetti opened the pull request:
An attempt at more general time series objects. Not taking for granted counts or count rates, having multiple array attributes, etc.
It might be, for example, spectroscopic or polarimetric light curves with multiple measurements per time point

Documentation for these changes can be found at https://github.com/StingraySoftware/notebooks/pull/73

~(Depends on #774)~ Merged

pep8speaks commented:
Hello @matteobachetti! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`stingray/base.py`](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py):

> [Line 77:21](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py#L77): [E713](https://duckduckgo.com/?q=pep8%20E713) test for membership should be 'not in'
> [Line 552:101](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py#L552): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 573:101](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py#L573): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 100 characters)
> [Line 578:101](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py#L578): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 604:101](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py#L604): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)
> [Line 655:101](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py#L655): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 100 characters)
> [Line 734:101](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py#L734): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 100 characters)
> [Line 1109:101](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py#L1109): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 1130:101](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py#L1130): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 100 characters)
> [Line 1135:101](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/base.py#L1135): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)











* In the file [`stingray/tests/test_base.py`](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/tests/test_base.py):

> [Line 311:101](https://github.com/StingraySoftware/stingray/blob/636a610342bfb0e1770d08939bc5e1e9a2fe1322/stingray/tests/test_base.py#L311): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +534,529 @@ def write(self, filename: str, fmt: str = None) -> None:
 
-        ts = self.to_astropy_table()
+        try:
+            Table({"a": np.longdouble(3)}).write(f".bu.{fmt}", overwrite=True)
```
```suggestion
            Table({"a": [np.longdouble(3)]}).write(f".bu.{fmt}", overwrite=True)
```

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +534,529 @@ def write(self, filename: str, fmt: str = None) -> None:
+            CAN_SAVE_LONGD = False
+
+        try:
+            Table({"a": 3}).write(f".bu.{fmt}", overwrite=True, serialize_meta=True)
```
```suggestion
            Table({"a": [3]}).write(f".bu.{fmt}", overwrite=True, serialize_meta=True)
```

matteobachetti commented:
Fixing things for making them work with the Generalized Light curves produced a monster PR. And there are more modifications to do. I will spin off the fixes to smaller PRs, and try to keep this PR on topic.

mgullik commented on stingray/crossspectrum.py:
```diff
@@ -1404,6 +1404,76 @@ def from_lightcurve(
+        use_common_mean=True,
+        gti=None,
+    ):
+        """Calculate AveragedCrossspectrum from two light curves
```
"from 2 time series"

mgullik commented on stingray/crossspectrum.py:
```diff
@@ -2080,6 +2150,88 @@ def crossspectrum_from_lightcurve(
+    gti=None,
+    save_all=False,
+):
+    """Calculate AveragedCrossspectrum from two light curves
```
"from two time series"

mgullik commented on stingray/crossspectrum.py:
```diff
@@ -1404,6 +1404,76 @@ def from_lightcurve(
 
+    @staticmethod
+    def from_stingray_timeseries(
+        lc1,
```
same in powerspectrum.py, 
the names of the variables (lc1 and lc2) can be confusing, but I don't know if it's worth changing them in ts1 and ts2

mgullik commented on stingray/crossspectrum.py:
```diff
@@ -2080,6 +2150,88 @@ def crossspectrum_from_lightcurve(
+
+
+def crossspectrum_from_timeseries(
+    lc1,
```
same comment of the previous method, name of the variables

mgullik reviewed (commented):
Hi @matteobachetti, 

this is an important change in the structure of the code. I will probably divide the review into a few steps

_join_timeseries and join method:
Can this method join StingrayTimeseries with EventList objects? 
If so, it might be useful to specify it (maybe I missed it) 
Does this method take into account when the MJDREF is present only in one or a portion of the objects that are going to be joint?

In lightcurve.py 
Is Lightcurve class a child class of StingrayTimeseries?
I don't see any change in the definition of the class Lightcurve, but I see a few methods that are deleted from here, I think this happened because Lightcurve is going to inherit all the methods of StingrayTimeseries, correct?

in crossspectrum.py and powerspectrum.py 
a few comments directly in the text about names 

I'm going to review the tests later

matteobachetti commented on stingray/base.py:
```diff
@@ -497,39 +1403,786 @@ def change_mjdref(self, new_mjdref: float) -> StingrayTimeseries:
+        in both. Otherwise, ``np.nan`` is used as a default value for the missing values.
+        Arbitrary array attributes are created and joined using the same convention.
+
+        Multiple checks are done on the joined time series. If the time array of the series
```
Done

matteobachetti commented on stingray/base.py:
```diff
@@ -497,39 +1403,786 @@ def change_mjdref(self, new_mjdref: float) -> StingrayTimeseries:
+
+        Parameters
+        ----------
+        other : :class:`EventList` object or class:`list` of :class:`EventList` objects
```
It is actually *not* possible to join different objects. This is a typo 😅

matteobachetti replied:
The type of the object must be the same. But note that the `join` method is also present in each of the other objects, with different docstrings.

matteobachetti commented on stingray/base.py:
```diff
@@ -497,39 +1403,786 @@ def change_mjdref(self, new_mjdref: float) -> StingrayTimeseries:
+
+        Returns
+        -------
+        lc_new: :class:`Lightcurve` object
```
Yep, leftover stuff. Sorry

matteobachetti commented on stingray/base.py:
```diff
@@ -497,39 +1403,786 @@ def change_mjdref(self, new_mjdref: float) -> StingrayTimeseries:
+
+    def rebin(self, dt_new=None, f=None, method="sum"):
+        """
+        Rebin the light curve to a new time resolution. While the new
```
Yeah, we do that in some places (e.g. `apply_mask`, `apply_gtis`), with the option `inplace`. It complicates the code flow quite a bit, and I'm not super sure it is useful in this case.

mgullik replied:
I agree!

matteobachetti commented:
@mgullik thanks for the review! I fixed all the docstrings. Indeed, many  of those methods were moved from `EventList` and `Lightcurve`, and the docstrings were left with the original class names.

In general, all child classes (`EventList`, `Lightcurve`) inherit the methods from `StingrayTimeseries` or re-write them if incompatible. `join` is partially re-written, but e.g. `EventList.join` does call `_join_timeseries`. However, note that there is a check that you _cannot_ join different types. If you start from an event list, you cannot merge it with a light curve, even if both of them are children of `StingrayTimeseries`.

Changes of MJDREF are accounted for. Note that you cannot be a StingrayTimeseries without defining MJDREF, at least as 0 (It is a field automatically defined in the `StingrayTimeseries` class, see line 1085).

mgullik commented on stingray/tests/test_base.py:
```diff
@@ -153,33 +329,651 @@ def test_hdf_roundtrip(self):
+            lc = lc0.concatenate(lc1)
+        assert lc.mjdref == 55000
+
+    def test_rebin(self):
```
I think rebin should be tested a bit stronger because it is a very important method. 
I would test it with a less-behaved time array. For example, with non-uniform arrival time. Also with multiple GTIs that simulate gaps in the timeseries

mgullik reviewed (commented):
Hi @matteobachetti 

this is the second part of the review. 
You find every comment on the code. 

I'm still not sure why StingrayTimeseries won't replace the EventList. 
If the time array in StingrayTimeseries doesn't have to be well-sampled, it is possible to lead an event file in StingrayTimeseries and then use rebin (dt = float) to produce/transform the StingrayTimeseries into a light curve. 
Of course, I don't want to suggest getting rid of EventList and Lightcurve, but I'd like to know if there is any functionality that EventList has and StingrayTimeseries hasn't

matteobachetti commented on stingray/tests/test_base.py:
```diff
@@ -84,6 +69,197 @@ def test_instantiate_without_main_array_attr(self):
+        ts1 = copy.deepcopy(self.sting_obj)
+        ts2 = copy.deepcopy(self.sting_obj)
+        # Add an array attribute to ts1. This will fail
+        ts1.blah = ts1.guefus * 2
```
Yes, it was just a way to give an attribute which is not linked in any way to existing attributes. Python assigns by pointer if you do `ts1.blah = ts2.guefus`, while it creates a new array if you do any operation on it. It's a minor thing, it does fail in any case.

mgullik replied:
got it!

matteobachetti commented on stingray/tests/test_base.py:
```diff
@@ -84,6 +69,197 @@ def test_instantiate_without_main_array_attr(self):
+        assert newts0.parafritus == "bonus!"
+        assert newts1.parafritus == "bonus!"
+        for obj in [newts1, newts0]:
+            assert obj.parafritus == "bonus!"
```
Yep 😅 . Fixed

matteobachetti commented on stingray/tests/test_base.py:
```diff
@@ -84,6 +69,197 @@ def test_instantiate_without_main_array_attr(self):
+    @pytest.mark.parametrize("inplace", [True, False])
+    def test_partial_apply_mask(self, inplace):
+        ts = copy.deepcopy(self.sting_obj)
+        obj = ts.apply_mask([True, True, False], inplace=inplace, filtered_attrs=["pardulas"])
```
This test is still on the base class `StingrayObject`, and in this particular example `guefus` is the main array attribute (like `time` in `Lightcurve` or `StingrayTimeseries`, or `freq` in `Crossspectrum`). So it will get filtered by default. `filtered_attrs` only lists the attributes besides the main array attribute

matteobachetti commented on stingray/tests/test_base.py:
```diff
@@ -84,6 +69,197 @@ def test_instantiate_without_main_array_attr(self):
+            assert np.array_equal(obj.pardulas, [3.0 + 1.0j, 2.0j])
+            assert np.array_equal(obj.sebadas, [[0, 1], [2, 3]])
+        assert ts is newts1
+        assert ts is not newts0
```
It should not fail, but in any case you are right that the logic is not very clear.
I made the test less confusing and prone to side effects by separating the two paths for inplace and not.

mgullik replied:
I think I understand now, 
the operand "is" and "is not" in python are not the same as doing == 
Is it correct that when you use inplace = True there is no new object and tsnew1 points to the same address of the original ts?

matteobachetti commented on stingray/tests/test_base.py:
```diff
@@ -153,33 +329,651 @@ def test_hdf_roundtrip(self):
+        assert "gti" not in array_attrs
+        assert "time" not in array_attrs
+
+    def test_operations(self):
```
Yes, in one case I was testing a very minimal subclass of the base `StingrayObj`, while here I'm testing that everything works on `StingrayTimeseries`, which is a quite more complex child with a lot more things that can go wrong 😅

matteobachetti commented on stingray/tests/test_base.py:
```diff
@@ -153,33 +329,651 @@ def test_hdf_roundtrip(self):
+        ts1 = StingrayTimeseries(time, array_attrs=dict(counts=count1), gti=gti1, dt=10)
+        ts2 = StingrayTimeseries(time, array_attrs=dict(counts=count2), gti=gti2, dt=10)
+        lc = ts1 - ts2
+        assert np.allclose(lc.counts, [300, 1100, 400])
```
Added a warning, non-overlapping data will be discarded.

matteobachetti commented on stingray/tests/test_base.py:
```diff
@@ -153,33 +329,651 @@ def test_hdf_roundtrip(self):
+        lc = -ts2 + ts1  # Test __neg__
+        assert np.allclose(lc.counts, [-300, -1100, -400])
+        assert np.array_equal(lc.time, time)
+        assert np.allclose(lc.counts_err, np.sqrt(2))
```
`test_failed_operations` tests exactly the case when the main array attribute is different (different dimension, different values). So, for `time` we are covered.
For all the rest, any array which is not of the same length as time will *not* be considered an array attribute, and thus will not be filtered/added/subtracted.

matteobachetti commented on stingray/tests/test_base.py:
```diff
@@ -153,33 +329,651 @@ def test_hdf_roundtrip(self):
+        lc1.change_mjdref(50001, inplace=True)
+        with pytest.warns(UserWarning, match="mjdref is different"):
+            lc = lc0.concatenate(lc1)
+        assert lc.mjdref == 55000
```
This is done in `test_concatenate_gtis_overlap`

mgullik reviewed (approved):
Hi @matteobachetti, 

thank you for addressing all my comments, I approve your Pull Request

dhuppenkothen commented on stingray/events.py:
```diff
@@ -246,14 +265,59 @@ def to_lc(self, dt, tstart=None, tseg=None):
+
+        The result will be something similar to a light curve, but with arbitrary
+        attributes corresponding to a weighted sum of each specified attribute of
+        the event list.
```
I'm not sure I understand here what you mean by "weighted" sum. If you have an EventList with an energy for each photon event, and you turn that into a `TimeSeries` object, should you not still have an array with time stamps and an array with energy values?

dhuppenkothen commented on stingray/events.py:
```diff
@@ -246,14 +265,59 @@ def to_lc(self, dt, tstart=None, tseg=None):
+        the event list.
+
+        E.g. if the event list has a ``q`` attribute, the final time series will
+        have a ``q`` attribute, which is the sum of all ``q`` values in each time bin.
```
Yeah, I still don't understand this. :D 
Does this mean that if you have a `q` that's multi-dimensional, say 3 values for every time bin, the resulting `TimeSeries` will have the sum of these three values for each bin in its attribute `q`? Why? Also, above it says "weighted sum". Weighted by what?

dhuppenkothen replied:
Ooooh, I see. It gets *binned*. But ... why? I thought `TimeSeries` could accommodate EventLists? Wasn't that the whole point of making it more general?

dhuppenkothen replied:
I think we need to think a bit carefully here how to structure this. Because right now, there are two separate behaviours supported by the same class: (1) You could make a TimeSeries object and just stick events into it, and that would work fine, or (2) you could generate a `TimeSeries` out of an EventList, but then it gets binned, so it's more like `Lightcurve`. I don't mind (2) intrinsically, but then we should not expose the user to the possibility of (1) even if we do that under the hood. I think I'd rather generalise `EventList` into something more general with more attributes if necessary, rather than allow the user to store EventLists as a time series object.

matteobachetti replied:
EventList is clearly a special case of a generic time series, and it makes sense to inherit much of the `StingrayTimeseries` class. On the other hand, creating a binned time series from an event list is useful.
I renamed the method as `to_binned_timeseries`, I think this is the easiest way to solve the conundrum. I have a time series object, I transform it into some other kind of time series object

dhuppenkothen replied:
Yeah, I think that's much better!

dhuppenkothen commented on stingray/events.py:
```diff
@@ -719,50 +666,6 @@ def filter_energy_range(self, energy_range, inplace=False, use_pi=False):
 
         return self.apply_mask(mask, inplace=inplace)
 
-    def apply_mask(self, mask, inplace=False):
```
So I'm assuming that this now inherits from the superclass? I'm wondering if it might still be worth having a custom docstring here for eventless specifically?

matteobachetti replied:
I don't know, it does exactly the same as `StingrayTimeseries`

dhuppenkothen replied:
Eh, it's a minor thing. I was wondering whether it confuses people if they look at a docstring for an EventList object and the docstring just talks about TimeSeries objects everywhere :) But maybe I'm not giving our users enough credit here. :D

dhuppenkothen commented on stingray/gti.py:
```diff
@@ -968,7 +968,7 @@ def join_equal_gti_boundaries(gti, threshold=0.0):
 
 
-def merge_gtis(gti_list, gti_treatment):
+def merge_gtis(gti_list, strategy):
```
Is this change to make the keyword argument consistent everywhere?

matteobachetti replied:
Yes, also.. I liked "strategy" more than "gti_treatment", it seemed more informative to what that argument does.

dhuppenkothen commented on stingray/lightcurve.py:
```diff
@@ -1628,10 +1479,10 @@ def from_lightkurve(lk, skip_checks=True):
 
-    def to_astropy_table(self):
-        return self._to_astropy_object(kind="table")
+    def to_astropy_table(self, **kwargs):
```
Needs docstring? Especially since it now takes `**kwargs` and the user might want to know what keyword arguments they can include. :)

dhuppenkothen commented on stingray/powerspectrum.py:
```diff
@@ -549,6 +549,63 @@ def from_events(
         )
 
+    @staticmethod
+    def from_stingray_timeseries(
```
same question as for Crossspectrum: does this enable the use of `Powerspectrum(ts1, ts2)` or does that still require only a light curve or event list?

matteobachetti replied:
Here too, I preferred not to add the complication of a new argument to `self.__init__`. But we could in principle allow that too

dhuppenkothen commented on stingray/tests/test_base.py:
```diff
@@ -47,43 +49,222 @@ def __init__(self, dummy=None):
+        # Add a non-scalar meta attribute to both, just slightly different
+        ts1.blah = [2]
+        ts2.blah = [3]
+        ts1 != ts2
```
should there be an assert statement here?

matteobachetti replied:
Oops

dhuppenkothen commented on stingray/tests/test_base.py:
```diff
@@ -47,43 +49,222 @@ def __init__(self, dummy=None):
+        assert np.array_equal(obj.guefus, [4, 5])
+        assert np.array_equal(obj.panesapa, ts.panesapa)
+        assert np.array_equal(obj.pardulas, [3.0 + 1.0j, 2.0j])
+        assert np.array_equal(obj.sebadas, [[0, 1], [2, 3]])
```
For my own entertainment, what are all those words, which I'm assuming are Italian? :D

matteobachetti replied:
They are all Sardinian desserts, in Sardinian language:
https://www.italiadelight.it/discover-your-favourite-traditional-sardinian-dessert/?lang=en

dhuppenkothen reviewed (changes requested):
Great work Matteo! I think this is a huge improvement! I have lots of minor nitpicks, a question about the relationship between EventList and TimeSeries, and a question for the future: with this added functionality, I'm expecting that people will want to somehow connect things like RMFs and ARFs to stingray. I know Matteo Lucchini is working on some code to do that. The question is: have you thought about how these would fit into the context of the current objects? Would they be separate objects? Would they get attached to a `TimeSeries` object as an attribute?

matteobachetti commented on stingray/base.py:
```diff
@@ -23,6 +25,94 @@
+]
+
+
+def _can_save_longdouble(probe_file: str, fmt: str) -> bool:
```
Done

matteobachetti commented on stingray/base.py:
```diff
@@ -23,6 +25,94 @@
+            probe_file, format=fmt, overwrite=True
+        )
+        yes_it_can = True
+        os.unlink(probe_file)
```
It should clean up, removing the test file

matteobachetti commented on stingray/base.py:
```diff
@@ -23,6 +25,94 @@
+
+def sqsum(array1, array2):
+    """Return the square root of the sum of the squares of two arrays."""
+    return np.sqrt(np.add(np.square(array1), np.square(array2)))
```
I figured it would be useful to have this function somewhere

matteobachetti commented on stingray/base.py:
```diff
@@ -61,36 +159,206 @@ def array_attrs(self) -> list[str]:
+
+        This also includes properties (which usually set internal data arrays, so they would
+        duplicate the effort), methods, and attributes that are complicated to serialize such
+        as other ``StingrayObject``, or arrays of objects.
```
I changed the description, hope it worrks

matteobachetti commented on stingray/base.py:
```diff
@@ -61,36 +159,206 @@ def array_attrs(self) -> list[str]:
         """List the names of the meta attributes of the Stingray Object.
 
         By array attributes, we mean the ones with a different size and shape
         than ``main_array_attr`` (e.g. ``time`` in ``EventList``)
```
No inputs, but I added the return values

matteobachetti commented on stingray/base.py:
```diff
@@ -61,36 +159,206 @@ def array_attrs(self) -> list[str]:
+
+    def dict(self) -> dict:
+        """Return a dictionary representation of the object."""
+        from collections import OrderedDict
```
Done

matteobachetti commented on stingray/base.py:
```diff
@@ -61,36 +159,206 @@ def array_attrs(self) -> list[str]:
+
+    def __str__(self) -> str:
+        """Return a string representation of the object."""
+        return self.__repr__()
```
They are not exactly the same. `pretty_print` can be customized with the input values. `__str__` is what is called when the user `print`s the data, and I figured it would be good to have an idea of the data contained in the object. 
For `__repr__`, I actually re-read the docs. I thought it was just a duplicate of `__str__`, but it's probably better to leave the default python does (`<StingrayObject> at blablahaddress`)

matteobachetti commented on stingray/base.py:
```diff
@@ -61,36 +159,206 @@ def array_attrs(self) -> list[str]:
+            return False
+
+        self_meta_attrs = self.meta_attrs()
+        other_meta_attrs = other_ts.meta_attrs()
```
Just to avoid calling the methods multiple times

matteobachetti commented on stingray/base.py:
```diff
@@ -61,36 +159,206 @@ def array_attrs(self) -> list[str]:
+
+        return True
+
+    def _default_operated_attrs(self):
```
Basically, the list of attributes that should be treated by default when summing etc. 
All array attributes, not including their error bars, that should be treated with the root of the sum of squares instead.

matteobachetti commented on stingray/base.py:
```diff
@@ -61,36 +159,206 @@ def array_attrs(self) -> list[str]:
+        operated_attrs = [attr for attr in self.array_attrs() if not attr.endswith("_err")]
+        return operated_attrs
+
+    def _default_error_attrs(self):
```
All error arttributes (e.g. counts_err)

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
             ts.write(filename, format=fmt, overwrite=True)
 
+    def apply_mask(self, mask: npt.ArrayLike, inplace: bool = False, filtered_attrs: list = None):
+        """Apply a mask to all array attributes of the time series
```
Right

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+            Array attributes to be filtered. Defaults to all array attributes if ``None``.
+            The other array attributes will be set to ``None``. The main array attr is always
+            included.
+
```
Yep

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+        """Operation that gets called with the ``+`` operator.
+
+        Add the array values of two time series element by element, assuming the ``time`` arrays
+        of the time series match exactly.
```
Yes, right, I created the methods for time series, then I generalized them but forgot to update the docstrings.

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+
+        This method implements overrides the ``len`` function for a :class:`StingrayTimeseries`
+        object and returns the length of the array attributes (using the main array attribute
+        as probe).
```
Just a bad docstring. This method is just what `len` uses to give a length.

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+        This method adds functionality to retrieve the count value at
+        a particular index. This also can be used for slicing and generating
+        a new :class:`StingrayTimeseries` object. GTIs are recalculated based on the new light
+        curve segment
```
Eeeeverything. It doesn't actually return the counts, but a new object with the slice of selected data.

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+    """Basic class for time series data.
+
+    This can be events, binned light curves, unevenly sampled light curves, etc. The only
+    requirement is that the data are associated with a time measurement.
```
Ok

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+        will run a check on the length of the arrays, and raise an error if they are not of a
+        shape compatible with the ``time`` array.
+
+    dt: float
```
It's in the "other parameters", absolutely optional, and it can be 0 or None for events. Also, it can be a single float for light curves, if uniformly sampled.
But it _might_ be useful if we are mixing up events with different frame times, for example.

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+        are not array attributes.
+
+    ncounts: int
+        The number of data points in the time series
```
Right, it's actually an attribute which was in `EventList` and is now deprecated

dhuppenkothen replied:
oh right, for `EventList` the number of data points would actually be the same as the total number of counts, but this is not true for time series generally. Is there a way we can be consistent with the rest of stingray? For example, in `Lightcurve` this is `Lightcurve.n`?

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+            setattr(self, kw, new_arr)
+
+        # if gti is None and self.time is not None and np.size(self.time) > 0:
+        #     self.gti = np.asarray([[self.time[0] - 0.5 * self.dt, self.time[-1] + 0.5 * self.dt]])
```
Done

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+            )
+        return value
+
+    @property
```
Done

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+    def apply_gtis(self, new_gti=None, inplace: bool = True):
+        """
+        Apply GTIs to a time series. Filters the ``time``, ``counts``,
+        ``countrate``, ``counts_err`` and ``countrate_err`` arrays for all bins
```
Right

matteobachetti commented on stingray/base.py:
```diff
@@ -497,39 +1475,791 @@ def change_mjdref(self, new_mjdref: float) -> StingrayTimeseries:
+        Concatenate two :class:`StingrayTimeseries` objects.
+
+        This method concatenates two or more :class:`StingrayTimeseries` objects. GTIs are
+        recalculated by merging all the GTIs together. GTIs should not overlap at any point.
```
Done

matteobachetti commented on stingray/base.py:
```diff
@@ -497,39 +1475,791 @@ def change_mjdref(self, new_mjdref: float) -> StingrayTimeseries:
+        return new_ts
+
+    def join(self, *args, **kwargs):
+        """
```
``concatenate`` is used to join multiple **non-overlapping** time series along the time axis, while ``join`` is more general, and can be used to join multiple time series with different strategies

matteobachetti replied:
I leave here the discussion about being consistent with pandas and astropy, although I'm not sure this is strictly necessary

dhuppenkothen replied:
I'm happy to not be consistent with them, as long as it's made clear somewhere in our docs what the difference between the two is :)

matteobachetti commented on stingray/crossspectrum.py:
```diff
@@ -1406,6 +1406,76 @@ def from_lightcurve(
             gti=gti,
         )
 
+    @staticmethod
```
The only problem with that is that the time series object does not have a default flux attribute, like lightcurve (which has `counts`). Using this method, you can specify the flux attribute. We could in principle add the flux keyword to `__init__`, but I'm not sure it's very helpful

dhuppenkothen replied:
Let's not worry about it for this version. I'll point out that one *can* make power spectra of all sorts of time series that aren't light curves, but I don't think it's worth worrying about it for now. It would be good to showcase this in the Powerspectrum tutorial to make sure users know to use a different syntax for TimeSeries than for LightCurve

matteobachetti commented on stingray/events.py:
```diff
@@ -88,7 +103,7 @@ class EventList(StingrayTimeseries):
 
     ncounts: int
-        Number of desired data points in event list.
+        Number of desired data points in event list. Deprecated
```
Done

matteobachetti commented on stingray/events.py:
```diff
@@ -246,14 +265,59 @@ def to_lc(self, dt, tstart=None, tseg=None):
-
         return Lightcurve.make_lightcurve(
-            self.time, dt, tstart=tstart, gti=self.gti, tseg=tseg, mjdref=self.mjdref
+            self.time, dt, tstart=tstart, gti=self._gti, tseg=tseg, mjdref=self.mjdref
```
It does. It's a property though: if gtis are not defined, they get created when the property is called by defining the internal attribute `self._gti`. In this part of the code, if gtis are not defined I can just pass self._gti, which is None, to lightcurve, and lightcurve will create the gtis once and for all after the data are binned.

matteobachetti commented on stingray/events.py:
```diff
@@ -20,25 +20,40 @@
+
+
 @njit
 def _from_lc_numba(times, counts, empty_times):
```
Done

matteobachetti commented on stingray/base.py:
```diff
@@ -390,21 +689,702 @@ def write(self, filename: str, fmt: str = None) -> None:
+        )
+
+    def _check_value_size(self, value, attr_name, compare_to_attr):
+        """Check if the size of a value is compatible with the size of another attribute.
```
I changed it to `_validate_and_format`... it should be less confusing, hopefully

matteobachetti commented:
@dhuppenkothen I should have addressed all your comments in one way or the other. 
About responses: they could definitely be added as attributes, and then used for whatever the user wants. Or we could just create another `Spectrum` object, maybe compatible with the equivalent `xspec` or `sherpa` objects. I think feedback from people like @matteolucchini1 , actually doing these things, should drive the choice though.

matteolucchini1 commented:
You can find the response stuff I've done so far here https://github.com/matteolucchini1/neXTsPec_prototype 
Currently what it does is simply take an impulse response in units of count rate as a function of energy and time (although the time axis could easily be e.g. pulse phase or Fourier frequency), and convolves that with response files to give units of count rate as a function instrument channel and time/whatever else.  In practice, what I actually chose to do is simply a matrix multiplication pretty much like x-psi, because it's much more straightforward than the xspec/isis/sherpa way of handling things and the computational cost is negligible. 

I'm not too familiar with Stingray, but in principle I think it should be possible to adapt my code to have a more robust energy:channel mapping, and then replace the rough_calibration function.

matteobachetti commented:
@matteolucchini1 thanks! We have a function that does something similar in HENDRICS: https://github.com/StingraySoftware/HENDRICS/blob/c0e2c68c965cd54404d7f1b859f1ad3bc96aa9f8/hendrics/calibrate.py#L133 
It's probable that yours will be more advanced, but feel free to take a look and recycle/propose changes at will
